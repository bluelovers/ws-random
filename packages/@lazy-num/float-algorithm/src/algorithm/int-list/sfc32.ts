import { MATH_POW_2_32 } from '../../const';

/**
 * Yet another chaotic PRNG, the sfc stands for "Small Fast Counter". It is part of the PracRand PRNG test suite. It passes PractRand, as well as Crush/BigCrush (TestU01). Also one of the fastest.
 */
export function df_sfc32(a: number, b: number, c: number, d: number)
{
	return function ()
	{
		a |= 0;
		b |= 0;
		c |= 0;
		d |= 0;
		const t = (a + b | 0) + d | 0;
		d = d + 1 | 0;
		a = b ^ b >>> 9;
		b = c + (c << 3) | 0;
		c = c << 21 | c >>> 11;
		c = c + t | 0;
		return (t >>> 0) / MATH_POW_2_32;
	};
}
