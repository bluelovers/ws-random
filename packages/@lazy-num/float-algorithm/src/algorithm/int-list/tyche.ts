import { MATH_POW_2_32 } from '../../const';

/**
 * Tyche is based on ChaCha's quarter-round. It's a bit slow but should be good quality. tychei, the inverted version, is 20% faster.
 */
export function df_tychei(a: number, b: number, c: number, d: number)
{
	return () =>
	{
		a |= 0;
		b |= 0;
		c |= 0;
		d |= 0;
		b = (b << 25 | b >>> 7) ^ c;
		c = c - d | 0;
		d = (d << 24 | d >>> 8) ^ a;
		a = a - b | 0;
		b = (b << 20 | b >>> 12) ^ c;
		c = c - d | 0;
		d = (d << 16 | d >>> 16) ^ a;
		a = a - b | 0;
		return (a >>> 0) / MATH_POW_2_32;
	};
}

/**
 * Tyche is based on ChaCha's quarter-round. It's a bit slow but should be good quality. tychei, the inverted version, is 20% faster.
 */
export function df_tyche(a: number, b: number, c: number, d: number)
{
	return () =>
	{
		a |= 0;
		b |= 0;
		c |= 0;
		d |= 0;
		a = a + b | 0;
		d = d ^ a;
		d = d << 16 | d >>> 16;
		c = c + d | 0;
		b = b ^ c;
		b = b << 12 | b >>> 20;
		a = a + b | 0;
		d = d ^ a;
		d = d << 8 | d >>> 24;
		c = c + d | 0;
		b = b ^ c;
		b = b << 7 | b >>> 25;
		return (b >>> 0) / MATH_POW_2_32;
	};
}
