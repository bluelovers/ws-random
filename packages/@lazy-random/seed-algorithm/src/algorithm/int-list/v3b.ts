/**
 * @example
 * var seed = 0; // any unsigned 32-bit integer
 * var next = v3b(seed, 2654435769, 1013904242, 3668340011);
 */
export function df_v3b(a: number, b?: number, c?: number, d?: number)
{
	b ||= 2654435769;
	c ||= 1013904242;
	d ||= 3668340011;

	let out: number, pos = 0, a0 = 0, b0 = b, c0 = c, d0 = d;
	return () =>
	{
		if (pos === 0)
		{
			a += d;
			a = a << 21 | a >>> 11;
			b = (b << 12 | b >>> 20) + c;
			c ^= a;
			d ^= b;
			a += d;
			a = a << 19 | a >>> 13;
			b = (b << 24 | b >>> 8) + c;
			c ^= a;
			d ^= b;
			a += d;
			a = a << 7 | a >>> 25;
			b = (b << 12 | b >>> 20) + c;
			c ^= a;
			d ^= b;
			a += d;
			a = a << 27 | a >>> 5;
			b = (b << 17 | b >>> 15) + c;
			c ^= a;
			d ^= b;

			a += a0;
			b += b0;
			c += c0;
			d += d0;
			a0++;
			pos = 4;
		}
		switch (--pos)
		{
			case 0:
				out = a;
				break;
			case 1:
				out = b;
				break;
			case 2:
				out = c;
				break;
			case 3:
				out = d;
				break;
		}
		return out >>> 0;
	};
}
