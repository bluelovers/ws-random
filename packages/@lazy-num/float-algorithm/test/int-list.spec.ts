import { df_tyche, df_tychei } from '../src/algorithm/int-list/tyche';
import { df_sfc32 } from '../src/algorithm/int-list/sfc32';

describe(`int-list`, () =>
{
	const actual_size = 16 as const;
	const actual_rest = [
		2654435769, 1013904242, 3668340011
	] as const;

	([
		df_tyche,
		df_tychei,
		df_sfc32,
	]).forEach(fn =>
	{

		test(fn.name, () =>
		{

			const seedFn: () => number = fn(0, ...actual_rest);

			let actual: number[] = [];

			for (let i = 0; i < actual_size; i++)
			{
				actual[i] = seedFn();
			}

			expect(actual).toMatchSnapshot();

			const seedFn2: () => number = fn(1, ...actual_rest);

			let actual2: number[] = [];

			for (let i = 0; i < actual_size; i++)
			{
				actual2[i] = seedFn2();
			}

			expect(actual).not.toStrictEqual(actual2);

			expect(actual2).toMatchSnapshot();

		});

	})

})
