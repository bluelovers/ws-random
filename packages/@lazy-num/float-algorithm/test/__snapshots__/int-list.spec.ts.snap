// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`int-list df_sfc32 1`] = `
Array [
  0.4721359545364976,
  0.5977381586562842,
  0.10499598179012537,
  0.7162471688352525,
  0.7737048789858818,
  0.10252661304548383,
  0.5388140785507858,
  0.03130675223655999,
  0.28249518550001085,
  0.9021577208768576,
  0.7259290926158428,
  0.715998571831733,
  0.4378080458845943,
  0.681514490628615,
  0.9879769335966557,
  0.22878749459050596,
]
`;

exports[`int-list df_sfc32 2`] = `
Array [
  0.47213595476932824,
  0.5977381586562842,
  0.10499598388560116,
  0.7206416979897767,
  0.7800575608853251,
  0.681634153239429,
  0.13451512763276696,
  0.11467337934300303,
  0.3707154141739011,
  0.05826257378794253,
  0.06896525109186769,
  0.709858814952895,
  0.8551409393548965,
  0.2190206153318286,
  0.927200733916834,
  0.446384334936738,
]
`;

exports[`int-list df_tyche 1`] = `
Array [
  0.23603756609372795,
  0.5645597272086889,
  0.7188761346042156,
  0.1841740373056382,
  0.36804007226601243,
  0.49389043520204723,
  0.33501260564662516,
  0.11455293255858123,
  0.15764436637982726,
  0.894220971269533,
  0.1071610082872212,
  0.5948217485565692,
  0.1850586417131126,
  0.31403911695815623,
  0.9815424710977823,
  0.1786510890815407,
]
`;

exports[`int-list df_tyche 2`] = `
Array [
  0.7379835366737098,
  0.7643441213294864,
  0.14950846857391298,
  0.515117404749617,
  0.16319397347979248,
  0.6185125396586955,
  0.3062659972347319,
  0.25807112478651106,
  0.7497057181317359,
  0.5236040477175266,
  0.8611736507155001,
  0.5043878313153982,
  0.7566755029838532,
  0.850998807232827,
  0.25944531220011413,
  0.9611672039609402,
]
`;

exports[`int-list df_tychei 1`] = `
Array [
  0.9641255775932223,
  0.6301178936846554,
  0.5457533597946167,
  0.3938262399751693,
  0.6295498560648412,
  0.17316253040917218,
  0.6257526748813689,
  0.3786617126315832,
  0.543666344601661,
  0.06360964197665453,
  0.4426391776651144,
  0.27776638767682016,
  0.5465525763574988,
  0.0884316407609731,
  0.49637938966043293,
  0.9143292678054422,
]
`;

exports[`int-list df_tychei 2`] = `
Array [
  0.9641255778260529,
  0.6298890369944274,
  0.32827600347809494,
  0.18519264529459178,
  0.7174751597922295,
  0.9828732563182712,
  0.6770932374056429,
  0.27858703047968447,
  0.20420499914325774,
  0.07651847042143345,
  0.681826448533684,
  0.4140391305554658,
  0.228399918647483,
  0.32219594344496727,
  0.1591027940157801,
  0.9879313134588301,
]
`;
